File: codes/jeongah/713_subarray_product_less_than_k.py
class Solution(object):
    def numSubarrayProductLessThanK(self, nums, k):
        count = 0
        end = 0
        product = 1
        
        for start in range(len(nums)):
            if end < start:
                end = start
                product = 1

            while end < len(nums):
                if product * nums[end] >= k:
                    count += end - start
                    break

                product *= nums[end]
                end += 1
            else:
                count += end - start

            product /= nums[start]

        return count

File: codes/jiheon/713_subarrayproductlessthank.java
class Solution {

    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k < 2)return 0;
        
        int l = 0, p = 1, ans = 0;

        for (int r = 0; r < nums.length; r++) {
            p *= nums[r];
            while (p >= k) {
                p /= nums[l++];
            }
            ans += r - l + 1;
        }
        
        return ans;
    }
}